### PR TITLE
[ARCTIC-394][Core]Support table watermark to determine table freshness

### LINKED ISSUE #394: [Feature]: Support table watermark to determine table freshness
### Description

Add table watermark property to determain table freshness

### Use case/motivation

Add a new table property `watermark` to telle table freshness

### Describe the solution

Here is a brief document about this feature:
https://docs.google.com/document/d/1LiwHPKBdQLKY3p9IyiKBI94K89B7brpfpbRnzp0k5Ts/edit#

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zhoujinsong**
## Why are the changes needed?

*Resolve #394*

## Brief change log

  - *Add watermark generator to calculate table watermark with new data files.*
  - *Calculate tablewatermark in add/overwrite/rowdelta/replacePartition operations*
  - *Remove watermark parameter from file cache*
  - *Add watermark configuration user documention*

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

File: hive/src/main/java/com/netease/arctic/hive/HiveTableProperties.java [hunk 1]
```diff
@@ -48,6 +48,8 @@ public class HiveTableProperties {
 
   public static final String ALLOW_HIVE_TABLE_EXISTED = "allow-hive-table-existed";
 
+  public static final String WATERMARK_HIVE = "watermark.hive";
```
Comment (hameizi) [inline hive/src/main/java/com/netease/arctic/hive/HiveTableProperties.java:51]: Is this unused?

Comment (zhoujinsong) [reply hive/src/main/java/com/netease/arctic/hive/HiveTableProperties.java:51]: Yes, both `watermark.base-store` and `watermark.hive` is unimplemented in this PR and will be done in future version. I would prefer to keep this property define here.

File: site/docs/ch/meta-service/table-properties.md [hunk 2]
```diff
@@ -68,6 +68,15 @@
 | log-store.data-version             | v1               | Logstore 中消息的版本，当前仅支持 v1           |
 | log.consistency.guarantee.enable   | false            | 标记是否开启一致性保证                       |
 
+### Watermark 相关配置
+
+| 配置名称                                       | 默认值               | 描述                                     |
+| ----------------------------------------------| ------------------- | ----------------------------------       |
+| table.event-time-field                        | _ingest_time        | 计算 watermark 的事件时间字段，默认的 _ingest_time 表示使用数据写入时间来计算 |
```
Comment (YesOrNo828) [inline site/docs/ch/meta-service/table-properties.md:75]: Should this field(`_ingest_time`) be exists in the table schema?

File: site/docs/ch/meta-service/table-properties.md [hunk 3]
```diff
@@ -68,6 +68,15 @@
...
+| table.watermark-allowed-lateness-second       | 0                   | 计算 watermark 时允许的数据乱序时间           |
+| table.event-time-field.datetime-string-format | yyyy-MM-dd HH:mm:ss | 当事件时间为字符串时，事件时间的格式           |
+| table.event-time-field.datetime-number-format | TIMESTAMP_MS | 当事件时间为数字时，事件时间的格式，支持 TIMESTAMP_MS(毫秒级时间戳)与TIMESTAMP_MS(秒级时间戳) |
```
Comment (YesOrNo828) [inline site/docs/ch/meta-service/table-properties.md:78]: TIMESTAMP_MS(毫秒级时间戳)与TIMESTAMP_MS(秒级时间戳) 
They are the same enums.

File: site/docs/ch/meta-service/table-management.md [hunk 4]
```diff
@@ -43,6 +45,16 @@ Logstore 作为可选组件则需要另外的配置来指明，有关 Logstore 
...
 而在执行上面的语句之前你还需要手动去对应的 Kafka 集群中创建对应的 topic，或者打开集群的自动创建 topic 功能。
 
+### 配置 Watermark
+水位（watermark） 用于表述表的写入进度，具体来说它是表上的一个类型为时间戳的属性，意为小于这个时间戳的数据都已经写入到表中，它一般用来观察表的写入进度，也可以作为下游批量计算任务的触发指标。
+
+上面的例子中将 `op_time` 设置为表的事件时间字段，在数据写入时会用写入数据的 `op_time` 来计算表的水位，同时为了应对写入乱序的问题，设置在计算水位时允许的数据迟到时间为一分钟。
+可以在 [AMS Dashboard](http://localhost:1630) 的表详情中看到表当前的水位，也可以在`Terminal`中输入下面的 SQL 来查询表的水位：
+
+```sql
+SHOW TBLPROPERTIES test_db.test_log_store ('table.watermark');
```
Comment (YesOrNo828) [inline site/docs/ch/meta-service/table-management.md:55]: Maybe showing the result of the table watermark in the Documentation is more intuitive and friendly for clients.

**Review by YesOrNo828 — COMMENTED** [review]
@zhoujinsong I left some comments.

File: site/docs/ch/meta-service/table-properties.md (hunk 2 above)
Comment (zhoujinsong) [reply site/docs/ch/meta-service/table-properties.md:75]: Not needed，`_ingest_time` is a a special configuration like the `process time` in Flink Job.

File: site/docs/ch/meta-service/table-properties.md (hunk 3 above)
Comment (zhoujinsong) [reply site/docs/ch/meta-service/table-properties.md]: Fixed.

File: site/docs/ch/meta-service/table-management.md (hunk 4 above)
Comment (zhoujinsong) [reply site/docs/ch/meta-service/table-management.md:55]: Added result part.

**Review by YesOrNo828 — APPROVED** [review]
LGTM
